lib/Backends/OpenCL/OpenCLDeviceManager.cpp

```diff
@@ -0,0 +1,197 @@
+  // Take the first platform.
+  cl_platform_id platform_id_used = platform_ids[0];
+  cl_uint num{0};
+  err = clGetDeviceIDs(platform_id_used, CL_DEVICE_TYPE_ALL, 0, nullptr, &num);
+  GLOW_ASSERT(err == CL_SUCCESS && "clGetDeviceIDs Failed.");
+  GLOW_ASSERT(num > deviceId &&
+              "Should have at least one GPU/CPU/FPGA for running OpenCL");
+  std::vector<cl_device_id> devices(num);
+  err = clGetDeviceIDs(platform_id_used, CL_DEVICE_TYPE_ALL, num,
+                       devices.data(), nullptr);
+  GLOW_ASSERT(err == CL_SUCCESS && "clGetDeviceIDs Failed.");
+  deviceId_ = devices[deviceId];
+  context_ = clCreateContext(nullptr, 1, &deviceId_, nullptr, nullptr, nullptr);
+  GLOW_ASSERT(context_ && "clCreateContext Failed.");
+  commands_ = clCreateCommandQueue(
+      context_, deviceId_, (doProfile_) ? CL_QUEUE_PROFILING_ENABLE : 0, &err);
+  GLOW_ASSERT(commands_ && "clCreateCommandQueue Failed.");
+  cl_ulong mem_size;
+  err = clGetDeviceInfo(deviceId_, CL_DEVICE_GLOBAL_MEM_SIZE, sizeof(cl_ulong),
+                        &mem_size, NULL);
+  GLOW_ASSERT(err == CL_SUCCESS && "Error getting device memory limit.");
+  maxMemoryBytes_ = mem_size;
+}
+
+OpenCLDeviceManager::~OpenCLDeviceManager() {
+  clReleaseCommandQueue(commands_);
+  clReleaseContext(context_);
+  buffers_.clear();
+}
+uint64_t OpenCLDeviceManager::getMaximumMemory() const {
+  return maxMemoryBytes_;
+}
+
+uint64_t OpenCLDeviceManager::getAvailableMemory() const {
+  return maxMemoryBytes_ - usedMemoryBytes_;
+}
+
+bool OpenCLDeviceManager::isMemoryAvailable(uint64_t estimate) const {
+  return maxMemoryBytes_ >= (usedMemoryBytes_ + estimate);
+}
+
+void OpenCLDeviceManager::addNetworkImpl(const Module *module,
+                                         FunctionMapTy functions,
+                                         ReadyCBTy readyCB) {
+  // First check for uniqueness of the function name.
+  for (const auto &func : functions) {
+    if (functions_.count(func.first) != 0) {
+      llvm::errs() << "Failed to add network: already have a function called "
+                   << func.first << ".\n";
+      readyCB(module, ResultCode::Failed);
+      return;
+    }
+  }
+  // Collect constants once, since currently the bundle grabs everything in the
+  // module.
+  auto bundle = functions.begin()->second->getRuntimeBundle();
+  bundle.collectConstants(module);
+  size_t sizeInBytes = bundle.getConstantWeightSize();
+  if (usedMemoryBytes_ + sizeInBytes > maxMemoryBytes_) {
+    llvm::errs() << "Failed to add network: not enough memory.\n";
+    // Free the constants.
+    bundle.freeConstants();
+    readyCB(module, ResultCode::Failed);
+    return;
+  }
+
+  // Copy constants to device.
+  auto size = bundle.getConstantWeightSize() + bundle.getMutableWeightSize() +
+              bundle.getActivationsSize();
+  auto deviceBuffer = allocDeviceBuffer(size);
+  auto buffer = std::make_shared<OpenCLBuffer>(deviceBuffer, size);
```

Comment:
We allocate the buffer rather early here, before the actual execution. And this reduces the amount of bytes available for the next network. If we add multiple networks this way, we may end-up with no memory available, even though none of them is being executed... Is it a concern? May be we should allocate on first use inside `runFunctionImpl`?

Reply:
We do this so we can copy the constants over to the device and release to Module. The plan is for this to be similar to the general device use case. It is definitely possible to run out of memory though. Hopefully the DeviceManager returns a failure to add network for OOM situations.